src/librbd/api/Namespace.cc

```diff
@@ -64,22 +64,36 @@ int Namespace<I>::create(librados::IoCtx& io_ctx, const std::string& name)
     return r;
   }
 
+  int ret_val;
   librados::IoCtx ns_ctx;
   ns_ctx.dup(io_ctx);
   ns_ctx.set_namespace(name);
 
+  r = ns_ctx.create(RBD_TRASH, false);
+  if (r < 0) {
+    lderr(cct) << "failed to create trash: " << cpp_strerror(r)
+               << dendl;
+    goto rollback_ns;
+  }
+
   r = cls_client::dir_state_set(&ns_ctx, RBD_DIRECTORY,
                                 cls::rbd::DIRECTORY_STATE_READY);
   if (r < 0) {
     lderr(cct) << "failed to initialize image directory: " << cpp_strerror(r)
                << dendl;
-    goto rollback;
+    goto rollback_trash;
   }
 
   return 0;
 
-rollback:
-  int ret_val = cls_client::namespace_remove(&default_ns_ctx, name);
+rollback_trash:
+  ret_val = ns_ctx.remove(RBD_TRASH);
+  if (ret_val < 0) {
+    lderr(cct) << "failed to remove trash: " << cpp_strerror(ret_val) << dendl;
```

Comment:
Will this lead to a leak if this fails but the ns is removed successfully?

Reply:
We're trying to rollback as best as we can. During rollback, if the removal of rbd_trash object fails, then the rbd_trash object will remain in the RADOS namespace. The RBD namespace creation will fail with return value of the operation that triggered the rollback. I'm not sure if we can do anything better?

Please note that cls_client::namespace_add() and cls_client::namespace_remove() only add or remove the RBD namespace names from the namespace directory object.  See, https://github.com/ceph/ceph/blob/v19.0.0/src/cls/rbd/cls_rbd.cc#L7865 and https://github.com/ceph/ceph/blob/v19.0.0/src/cls/rbd/cls_rbd.cc#L7902

A RBD namespace is a RADOS namespace with metadata set on rbd_namespace (namespace directory) and rbd_directory (image directory) objects. Please see https://lists.ceph.io/hyperkitty/list/ceph-users@ceph.io/thread/PEDRYSNNDZ7B6FLMKVBNUMOYE6EK2H4P/ , especially Paul's reply in the thread, https://lists.ceph.io/hyperkitty/list/ceph-users@ceph.io/message/T2YK77HDWBNT5ADACRERU7EFYZCN3K5P/ 
As Paul says, you can think of a RADOS namespace as a prefix that gets added to an object.